Alex' updates

scpalbright opened the pull request.

alasheen committed:
Correcting some comments in Section examples

alasheen committed:
Starting with some simple examples with Ring

alasheen committed:
Moving the original input as a classmethod

alasheen committed:
Intermediate push with from_simple_input classmethod

alasheen committed:
Test implementation with classmethod and buffer to pipe data to __init__

Commit kept for reference in the eventuality where the RFStation default
signature should be RFStation() with no optional arguments as baseline
for the empty RFStation

alasheen committed:
Buffer using dict instead of obj

alasheen committed:
Merge pull request #16 from scpalbright/future

Align with scpalbright future branch

alasheen committed:
Adding sphinx-autopackagesummary to travis

alasheen committed:
Adapting the criterion to identify equal maxima (based on minimum)

alasheen committed:
Adding perftest and unittest to find_potential_wells_cubic

alasheen committed:
Merge pull request #17 from scpalbright/future

Getting latest changes from scpalbright

alasheen committed:
Section with new datatype structure

alasheen committed:
Adding length_function to datatype.ring_program (section combination)

alasheen committed:
Adding fix to get proper time base in combine_single_sections

A better solution will be found, this is a workaround for now

alasheen committed:
Removing print

alasheen committed:
Travis: don't build the doc if pull request (env variable are missing)

alasheen committed:
TRAVIS_PULL_REQUEST is an int, checking if int to exit

alasheen committed:
Renaming Section->RingSection and section_length->length

alasheen committed:
Adding orbit_length as program instead of length

Unittests to check orbit_length as program to follow

alasheen committed:
Adding unittest for orbit_length as program

alasheen committed:
Adding small unittest for coverage

alasheen committed:
Unittest for datatype as input to RingSection for synchronous data

alasheen committed:
Adding test_datatype_input and rename test_machine_program_input

alasheen committed:
Adding checks that orbit_length size is in line with sync_data

Unittests to follow

alasheen committed:
Checking TRAVIS_PULL_REQUEST env variable

alasheen committed:
Adjusting test TRAVIS_PULL_REQUEST

alasheen committed:
Adding warning if sync_data is single and other input progs + tests

alasheen committed:
Renaming orbit_length to orbit_length_program in dtype

alasheen committed:
Forgot some propagation of orbit_length_program change

alasheen committed:
Ring working with the RingSection structure. WIP!!

Documentation to be updated, full coverage with unittest to be achieved,
direct_input classmethod with datatypes to be tested

alasheen committed:
Adding unittest for dtype input to RingSection

alasheen committed:
Renaming ring_options, will be removed when all func re-implemented

alasheen committed:
Renaming test_section to test_ring_section

alasheen committed:
Removing print and 1st Ring unittest

alasheen committed:
Improved declaration of undefined alpha with dtypes

alasheen committed:
Renaming orbit_length opt to orbit_bump + length/length_d attributes

alasheen committed:
Commenting Ring unittest for now

alasheen committed:
Renaming alpha_order to alpha_orders (now a list)

alasheen committed:
Adding minor doc update

alasheen committed:
Fix in the example (forgotten renaming)

alasheen committed:
Corrected Ring object with modifications from RingSection

alasheen committed:
Renaming input_parameters to machine_parameters

alasheen committed:
Correcting imports after renaming

alasheen committed:
Added design (and including orbit bumps) for various Ring attributes

alasheen committed:
Adding eta_orders as an option

alasheen committed:
Updated docstrings examples

alasheen committed:
Adding conversion of synchronous_data in Ring + some examples

alasheen committed:
Adding Ring with turn/time base examples, correction to dtype.preprocess

Definition of n_turns to be checked

alasheen committed:
Adding example with orbit_bump time based prog + correction to Ring

alasheen committed:
PEP8 ring_programs

alasheen committed:
Adjusting multisection turn based program in preprocess

alasheen committed:
Correcting index for multi section calculation

alasheen committed:
Reorganization of the momentum extraction, virtually identical

alasheen committed:
Stopping the preprocess function before the last input time value

alasheen committed:
Renaming old examples to be deleted when new version complete

alasheen committed:
Adding an example using machine_program

alasheen committed:
Simplifying the input check of Section_list

scpalbright commented on datatypes/ring_programs.py:
```diff
@@ -1,32 +1,40 @@
 from ._core import _function, _expand_function, _check_time_turns,\
-                   _get_dats_types, _check_data_types, _expand_singletons,\
-                   _check_turn_numbers
+    _get_dats_types, _check_data_types, _expand_singletons,\
+    _check_turn_numbers
```
Is this indentation the style guide version?  I don't like it...

alasheen replied:
Would this be better? (with all the imports aligned, without or without the parenthesis)
```python
from ._core import (_function, _expand_function, _check_time_turns,
                    _get_dats_types, _check_data_types, _expand_singletons,
                    _check_turn_numbers)
```

scpalbright replied:
I find it clearer and a bit more aesthetically pleasing, but it's not exactly a major thing.

alasheen replied:
Corrected

scpalbright commented on datatypes/ring_programs.py:
```diff
@@ -500,69 +505,73 @@ def preprocess(self, mass, circumference, interp_time = None,
+
             if t_end > self[0, 0, -1]:
-                warnings.warn("t_stop too late, ending at " 
-                              + str(self[0, 0, -1]))
+                if not np.isinf(t_end):
```
No need for double `if`, could just be:
```python
if t_end > self[0, 0, 0] and not np.isinf(t_end):
```

alasheen replied:
You'd still need a second if to catch the case where t_end is np.inf?

scpalbright replied:
Ah, misread the indentation.  All good.

scpalbright commented on datatypes/ring_programs.py:
```diff
@@ -776,10 +781,9 @@ def _time_from_turn(self, mass, circumference):
         trev = rt.mom_to_trev(self[0], mass, circ=circumference)
-        return np.cumsum(trev)
-
+        return np.insert(np.cumsum(trev[:-1]), 0, 0)
```
Is this needed for the `n_turns+1` thing?

alasheen replied:
No that's to avoid getting a t_rev offset on the cycle_time (passing a program by_time would start from 0 or t_start while by_turn would start from t_rev)

scpalbright replied:
Ooh, I see.  Why not just do:
```python
time = np.cumsum(trev)
return time-time[0]
```
?

alasheen replied:
Definitely an option! I am wondering for very large arrays what is the most efficient. This can come at a second stage when we can evaluate how long does it take to generate a Ring object for our default use cases (although I believe this part won't be that critical)

scpalbright replied:
Definitely slower to use the insert, but not enough to matter I think.
```python
In [38]: test = np.linspace(0, 1, 10000)

In [39]: %timeit np.cumsum(test)
37.2 µs ± 58.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [40]: %timeit np.cumsum(test) - 0.1
49 µs ± 146 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [41]: %timeit np.insert(np.cumsum(test[:-1]), 0, 0)
82.7 µs ± 178 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

I think in this instance speed isn't really a concern, I'm just wondering if missing the last point could change the result

alasheen replied:
Ah yes you reminded me, one thing we need to consider is indexing, to keep
cycle_time[i] = cycle_time[i-1] + t_rev[i-1] with i>0

Like this np.diff(cycle_time) == t_rev (with the last element missing) aligned with np.diff(energy) == delta_E

(I hope I didn't get the index wrong 😄)

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,588 @@
+        # based on the name
+        if isinstance(Particle, beam.Particle):
+            self.Particle = Particle
+        elif isinstance(Particle, str):
+            self.Particle = beam.make_particle(Particle)
+        else:
+            raise excpt.InputError("The input Particle should be a " +
+                                   "Particle object or a str.")
```
As mentioned in Part1 pull request

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,588 @@
+
+        # Storing the momentum program and computing all associated values
+        self.momentum = momentum_processed[2:]
+        self.beta = rt.mom_to_beta(
+            momentum_processed[2:], self.Particle.mass)
+        self.gamma = rt.mom_to_gamma(
+            momentum_processed[2:], self.Particle.mass)
+        self.energy = rt.momentum_to_energy(
+            momentum_processed[2:], self.Particle.mass)
+        self.kin_energy = rt.momentum_to_kin_energy(
+            momentum_processed[2:], self.Particle.mass)
```
Needs `beta` and `gamma` subclasses of `_synchronous_data_program` adding, then these can all be done as:
```python
self.beta = self.momentum.to_beta(False, mass = self.Particle.mass)
# etc
```
As written this will end up with `self.energy` being of type `momentum_program`, which could cause confusion....

alasheen replied:
Corrected

For beta,gamma I cast those as np.arrays in the meanwhile to avoid confusion when printing

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,588 @@
+        self.omega_rev = 2 * np.pi * self.f_rev
+
+        # Recalculating the delta_E
+        self.delta_E = np.diff(self.energy, axis=1)
+        if self.n_turns > len(self.use_turns):
+            self.delta_E = np.zeros(self.energy.shape)
+            self._recalc_delta_E()
```
Looks a little awkward to me, maybe better as:
```suggestion
        if self.n_turns > len(self.use_turns):
            self.delta_E = np.zeros(self.energy.shape)
            self._recalc_delta_E()
        else:
            self.delta_E = np.diff(self.energy, axis=1)
```

alasheen replied:
Corrected

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,588 @@
+
+            for section in self.Section_list:
+
+                if hasattr(section, alpha_name):
+                    alpha_prog.append(getattr(section, alpha_name))
+                else:
+                    alpha_prog.append(ring_programs.momentum_compaction(
+                        0, order=order))
```
"Easier to ask for forgiveness than permission"

```suggestion
                try:
                    alpha_prog.append(getattr(section, alpha_name))
                except AttributeError:
                    alpha_prog.append(ring_programs.momentum_compaction(
                        0, order=order))
```

`hasattr` effectively does a `getattr` and returns `True` if there's no `AttributeError` and `False` if there is an `AttributeError`.  May as well do it ourselves.

scpalbright replied:
Side note:
It's faster to use `try/except` if there is no exception raised, but faster to use `if/else` if the condition is `False`.  Obviously this is a single call so it's not really relevant, but if something is run repeatedly and the `exception` is expected to be rare, it's faster to use `try/except` than `if/else`.

alasheen replied:
Corrected

Thanks for the side note I was about to ask that specifically!

scpalbright commented on __TESTS/perftests/rf_functions/test_findpotwells.py:
```diff
@@ -0,0 +1,175 @@
+        self.time_array, self.rf_voltage_array = rf_voltage_generation(
+            n_points, t_rev, voltage, harmonic, phi_rf, time_bounds=time_bounds)
+    
+        from blond_common.rf_functions.potential import rf_potential_generation
```
```suggestion

```

scpalbright commented on __TESTS/perftests/rf_functions/test_findpotwells.py:
```diff
@@ -0,0 +1,175 @@
+    
+        rf_station = RFStation(ring, harmonic, voltage, phi_rf)
+    
+        from blond_common.rf_functions.potential import rf_voltage_generation
```
```suggestion

```

scpalbright commented on __TESTS/perftests/rf_functions/test_findpotwells.py:
```diff
@@ -0,0 +1,175 @@
+        ring = Ring(ring_length, alpha_0, particle, bending_field=bending_field,
+                    bending_radius=bending_radius)
+    
+        from blond_common.interfaces.input_parameters.rf_parameters import RFStation
```
```suggestion

```

scpalbright commented on __TESTS/perftests/rf_functions/test_findpotwells.py:
```diff
@@ -0,0 +1,175 @@
+        
+        # Defining a ramp with a program time vs. energy (warning: initial energy cannot be 0)
+    
+        from blond_common.interfaces.beam.beam import Particle
+        from blond_common.interfaces.input_parameters.ring import Ring
+        from scipy.constants import u, c, e
```
```suggestion

```

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,588 @@
+        # Casting synchronous data to datatype
+        if not isinstance(synchronous_data,
+                          ring_programs._synchronous_data_program):
+            synchronous_data \
+                = ring_programs._synchronous_data_program._conversions[
+                    func_type](synchronous_data)
```
```suggestion
            synchronous_data  = getattr(ring_programs, func_type)(synchronous_data)
```

alasheen replied:
Corrected with renaming of local syncDataType

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,588 @@
+                **kwargs)
+
+            Section_list.append(section)
+
```
Most of the data checking is a repeat of what is done in the `RingSection` objects.  Should be easy to just get everything into a list of the correct length and then construct all the sections, and let them handle the checking of stuff.  Something like:
```python
sync_data = {k, v for k, v in zip(syncDataTypes, syncDataInput)}
for k in sync_data:
    if sync_data[k] is None:
        sync_data[k] = [None]*n_sections

alphas = {'alpha_0': alpha_0, 'alpha_1': alpha_1, 'alpha_2': alpha_2, 
                **{k: v for k, v in kwargs if 'alpha_' in k}}
for k in alphas:
    if alphas[k] is None:
        alphas[k] = [None]*n_sections
```
Then a quick length check:
```python
if not all([len(v) == n_sections for v in {**sync_data, **alphas}.values()])
    raise SomeError("Your data is crap")
```
Then the section can be made as:
```python
for i in range(n_sections):
    sectionInput = {**{k: sync_data[k][i] for k in sync_data},
                             **{k: alphas[k][i] for k in alphas}, **kwargs}
    Section_list.append(RingSection(**sectionInput))
```

I think something like this would reduce the amount of duplicate input testing and be easier to maintain if additional options are added.

alasheen replied:
Noted, the direct_input is still work in progress I'll keep these in mind because indeed we need to minimize redundant code

alasheen committed:
Simplifying call to private member

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,588 @@
+                                            self.omega_rev)
+        parameters['eta_0'] = np.interp(cycle_moments, self.cycle_time,
+                                        self.eta_0[0])
+        if len(self.cycle_time) == len(self.delta_E[0]):
+            parameters['delta_E'] = np.interp(cycle_moments,
+                                              self.cycle_time,
+                                              self.delta_E[0])
+        else:
+            parameters['delta_E'] = np.interp(cycle_moments,
+                                              self.cycle_time[:-1],
+                                              self.delta_E[0])
```
I really don't like this, but couldn't think of a better solution.  If you can think of one I'll be grateful...

alasheen replied:
I think this came from the fact that the definition of n_turns was different if the data was passed by_turn or by_time in preprocess, now that all programs should all be in size n_turns+1 except delta_E which is of size n_turns, we can keep only the 'else' statement

alasheen replied:
What about having kwargs to select which attributes to get from this function?

scpalbright replied:
No it's the length thing with `n_turns` or `n_turns+1`.  I did it, I just don't like it!

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,588 @@
+
+        return parameters
+
+    def _recalc_delta_E(self):
+        """
+        Function to recalculate delta_E.
+        If interpolation is not done on a turn-by-turn basis the delta_E will
+        not be correct.  This function recalculates it to give the correct
+        value for each turn.
+        """
+
+        for section in range(self.n_sections):
+            derivative = calc.deriv_cubic(self.cycle_time,
+                                          self.energy[section])[1]
+            self.delta_E[section] = derivative * self.t_rev
```
This function does odd things when there's few points and a transition from flat bottom to ramp.  I can't find one that resolves the ramp well and also does the flat bottom well.  Know of anything better?

alasheen replied:
I'd avoid cubic_spline here, especially for cases where Bdot is defined as piecewise linear functions, maybe worth looking into the linear functions instead but not sure if gradient or a modified version of diff (we would have the issue that the output is one element short) is best
.

scpalbright replied:
Yeah, I tried gradient and diff, but they give different problems elsewhere.  This gave the best results on the ramp, but then the flat bottom went wrong.  Never mind, I'll worry about it another time.

scpalbright commented on interfaces/machine_parameters/ring_section.py:
```diff
@@ -177,98 +165,134 @@ def __init__(self, section_length, alpha_0,
-                    self.alpha_order = np.max([self.alpha_order, order])
+                    alpha_order_max = np.max([alpha_order_max, order])
                     alpha_n[order] = kwargs[argument]
                 except Exception:
```
Eeeesh.  `except Exception` is risky...

alasheen replied:
Replaced Exception by ValueError (the goal here was to catch wrong casting to`int(argument.split('_')[-1])`, all cases I can think of lead to ValueError but I set to Exception originally for cases I wouldn't have considered)

scpalbright replied:
You're right, I tested with impossible input, since they're always coming from kwargs it'll never be anything but `str`.  If a `TypeError` ever comes up there's a MUCH bigger problem than a bad kwarg.

scpalbright commented on interfaces/machine_parameters/ring_section.py:
```diff
@@ -177,98 +165,134 @@ def __init__(self, section_length, alpha_0,
 
         # Treating non-linear momentum compaction factor if declared
         # Listing all the declared alpha first
         alpha_n = {1: alpha_1, 2: alpha_2}
```
You can change this bit, add the exception handling at the same time and remove the need for the loops below.
```suggestion
        try:
            alpha_n = {1: alpha_1, 2: alpha_2,
                             **{k.split('_')[-1]:kwargs[k] for k in kwargs if 'alpha_' in k}}
        except (TypeError, ValueError):
            # Do some stuff to figure out what went wrong
            raise StupidUserError("User, you are stupid")
        
        alpha_order_max = np.max(alpha_n.keys())
```

alasheen replied:
Impressive! I think I'll keep the \_\_init\_\_ a bit more verbose for now in case a new dev in the future tries to read through, it will be easier, but for the `direct_input` method since this is in line with the suggestions few comments above I'll give it a try

scpalbright replied:
You do over long names, I do over dense code!

alasheen committed:
Undo previous commit

scpalbright commented on interfaces/machine_parameters/ring_section.py:
```diff
@@ -177,98 +165,134 @@ def __init__(self, section_length, alpha_0,
-        self.alpha_orders_defined = [0]
-        for order in range(1, self.alpha_order+1):
+        for order in range(1, alpha_order_max + 1):
             alpha = alpha_n.pop(order, None)
 
             if alpha is None:
                 alpha = 0
```
```suggestion
            alpha = alpha_n.pop(order, 0)
```

alasheen replied:
This if statement is placeholder, we can replace `alpha = 0` with `continue` to avoid filling undeclared alpha with 0 values

(1M dollar question, what would `alpha = alpha_n.pop(order, continue)` do?)

scpalbright replied:
Ah ok, in that case I'd do:
```python
try:
    alpha = alpha_n.pop(order)
except KeyError:
    continue
```

And it'll do:
```
>>> {1:1}.pop(1, continue)
  File "<stdin>", line 1
    {1:1}.pop(1, continue)
                        ^
SyntaxError: invalid syntax
```

alasheen replied:
Included

alasheen committed:
Corrected wrong units

alasheen committed:
Corrected imports + call to test, need to be fixed with RFStation

alasheen committed:
Renaming Section_list to RingSection_list + docs

alasheen committed:
Docs correction

alasheen committed:
Added orbit_bump as attribute to RingSection

alasheen committed:
Fixing RingSection unittests

alasheen committed:
Small docs addition

alasheen committed:
Correcting call to beam.make_particle

alasheen committed:
Correction in comments

alasheen committed:
More beauty for import

alasheen committed:
Avoiding copy() function and casting cycle_time as np.array

scpalbright commented on interfaces/machine_parameters/ring_section.py:
```diff
@@ -170,31 +171,37 @@ def __init__(self, length, alpha_0,
             if orbit_bump.timebase == 'by_time':
-                self.length[:, 1, :] += self.length_design
+
+                self.length = ring_programs.orbit_length_program(
+                    [self.length_design] * len(self.orbit_bump[0, 0, :]),
+                    time=self.orbit_bump[0, 0, :])
```
The shaping will be handled internally by the `datatype` object if you pass `time` as an argument.
```suggestion
                self.length = ring_programs.orbit_length_program(
                    self.length_design,
                    time=self.orbit_bump[0, 0, :])
```

alasheen replied:
I tried but unfortunately it gave me a `single` datatype.

In a test
```python
>>> ring_programs.orbit_length_program(1., time=np.array([0,1,2]))
orbit_length_program([1.])
```

scpalbright replied:
Shit.

Could you add that to the improvements issue?

alasheen replied:
Added in the github issue

alasheen committed:
Chaning conversion of beta,gamma,energy,kin_energy

alasheen committed:
Better if/else for self.delta_E

alasheen committed:
Replacing hasattr by try/except AttributeError

alasheen committed:
Replace Exception by ValueError

alasheen committed:
Improving prints in RingSection examples

alasheen committed:
Adding comment on RingSection examples

scpalbright commented on __EXAMPLES/interfaces/ring_section.py:
```diff
@@ -16,7 +16,7 @@
 
 # BLonD Common import
-sys.path.append('./../../../')
+sys.path.append('./../../../')  # Add blond_common to PYTHONPATH if not done
```
```suggestion
sys.path.append('../../../')  # Add blond_common to PYTHONPATH if not done
```

Personally I wouldn't do this at all.  This will only run in the correct directory, it will be an issue if someone copies it to another directory and tries to run it.

alasheen replied:
Removed path manip in example

alasheen committed:
try/except instead of if, only alpha_1 as explicit kwarg

alasheen committed:
Remove sys.path manipulation in example

alasheen committed:
Raising warning with __ring_options import

alasheen committed:
Improved Ring.direct_input, all should be handled but tests needed

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,598 @@
+
+                    else:
+
+                        raise excpt.InputError(
```
Perhaps a better exception would be `NotImplementedError`.  The input here is fine, we just haven't yet figured out how to treat it.
```suggestion
                        raise NotImplementedError(
```

alasheen committed:
Adding missing warning for corner case

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,598 @@
+        for order in range(self.eta_orders + 1):
+            setattr(self, 'eta_%d' % (order), np.zeros(self.momentum.shape))
+        self._eta_generation()
+
```
By this point the kwargs should be empty.  Maybe worth putting something like:
```python
if len(kwargs) > 0:
    warings.warn(f"Unused kwargs have been detected, they are {list(kwargs.keys())}")
```

alasheen replied:
Added in both Ring and RingSection, this can be generalized everywhere I guess

scpalbright replied:
Depends how they're used.  If we have a specific set of required kwargs and we pop them all out then we can be sure it's empty by the end.  If they get passed on to other functions and objects we can't know that.  Where it's possible I think we should aim to include it though.

alasheen committed:
Replace exception name

Co-authored-by: scpalbright <scpalbright@users.noreply.github.com>

alasheen committed:
Adding warnings for unused kwargs

alasheen committed:
Merge branch 'future' of https://github.com/alasheen/blond_common.git into future

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -0,0 +1,604 @@
+
+            sectionInput = {'length': length[i],
+                            'bending_radius': bending_radius,
+                            **{k: sync_data[k][i] for k in sync_data},
+                            **{k: alphas[k][i] for k in alphas},
```
Short term solution to the casting issue:
```suggestion
                            **{k: sync_data[k][i].view(np.ndarray) for k in sync_data},
                            **{k: alphas[k][i].view(np.ndarray) for k in alphas},
```

alasheen replied:
Included, I just added an if statement for None cases (keeping the inline coding for maximum compression rate 😄 )

scpalbright replied:
Hard approve!!!

alasheen committed:
Fix for Ring.direct_input

The datatypes are casted as np.arrays for now before being passed to
RingSection (timebase definition)

scpalbright commented on interfaces/machine_parameters/ring_section.py:
```diff
@@ -177,98 +168,162 @@ def __init__(self, section_length, alpha_0,
-                alpha = 0
+            try:
+                alpha = alpha_n.pop(order)
+                self.alpha_orders.append(order)
```
I would move this line outside the `try/except` block.  It's best practice to put the bare minimum in the `try`, you can't generate a `KeyError` with it but still best not to have it there I think.  You could put it in an `else` block after the `except` block to make it explicit that it's related to what's happening in the `try` block.

alasheen replied:
Included

scpalbright commented on interfaces/machine_parameters/ring_section.py:
```diff
@@ -177,98 +168,162 @@ def __init__(self, section_length, alpha_0,
                 try:
                     order = int(argument.split('_')[-1])
-                    self.alpha_order = np.max([self.alpha_order, order])
+                    alpha_order_max = np.max([alpha_order_max, order])
                     alpha_n[order] = kwargs[argument]
```
As with the one below, better to put these two outside the `try/except`, an `else` block would work but is not necessary.

alasheen replied:
Included with the above

alasheen committed:
Keeping only necessary code in try/except

alasheen committed:
Adding _Ring_opt_dflt as Ring options default values

alasheen committed:
Renaming legacy module of datatypes, all func moved to dedicated modules

alasheen committed:
Correcting source of dTypes function in machine_program

alasheen committed:
Renaming all deprecated files with __

alasheen committed:
Adding -- to help viewing print for different cases

scpalbright commented on interfaces/machine_parameters/ring.py:
```diff
@@ -29,6 +29,16 @@
+_Ring_opt_dflt['interpolation'] = 'linear'
+_Ring_opt_dflt['store_turns'] = True
+_Ring_opt_dflt['eta_orders'] = 0
+
```
Maybe worth adding a little function to print the defaults?
```suggestion
def RingKeywordDefaults():
    for k in _Ring_opt_dflt:
        print(f"For keyword {k}, the default value is: {_Ring_opt_dflt[k]}")
```
What do you think?

alasheen replied:
It's a good idea. I think we could have a more general approach since many if not all the modules will have default option values. Maybe let's keep the utilities to display and modify default option values for a specific discussion?

scpalbright commented on interfaces/machine_parameters/ring_section.py:
```diff
@@ -227,31 +227,33 @@ def __init__(self, length, alpha_0,
                         'factor. ' +
                         'The correct syntax is alpha_n.')
 
+                alpha_order_max = np.max([alpha_order_max, order])
+                alpha_n[order] = kwargs.pop(argument)
```
Wouldn't make much difference, but you could avoid the repeated `np.max` call by waiting until the end of the `for` loop and doing:
```python
alpha_order_max = np.max(list(alpha_n.keys()))
```

alasheen replied:
Almost included

I have to catch the case where no alpha_n is declared. In that case I'd go for an if/else since the majority of the time no alpha_n is declared

```python
        if len(list(alpha_n.keys())) > 0:
            alpha_order_max = np.max(list(alpha_n.keys()))
        else:
            alpha_order_max = 0
```

would you still prefer a try/except?

scpalbright replied:
Ah, you don't include `alpha_0` as part of the `alpha_n` dictionary.  In that case I'd do:
```python
alpha_order_max = np.max(list(alpha_n.keys()) + [0])
```

You can also remove the two declarations of it as 0 and 1 on lines 211 and 228

alasheen replied:
Included!

scpalbright replied:
Side note 2:  You can call `len` on a `dict`, it tells you the number of elements it contains.  So you will get identical result for `len(alpha_n)` as for `len(list(alpha_n.keys()))`.

alasheen committed:
Correction in direct_input, syncDataTypes and syncDataArgs not identical

alasheen committed:
Correcting bending_radius treatment in direct_input

alasheen committed:
Avoiding repetition of max calculation

alasheen committed:
Removing useless early definition of alpha_order_max

alasheen committed:
Including orbit_bump in Ring

alasheen committed:
Tidying up all Ring examples and improving prints

scpalbright commented on datatypes/ring_programs.py:
```diff
@@ -130,10 +136,10 @@ def _combine_single_sections(cls, *args,
-            assrt.equal_arrays(*timeBases, msg = 'Attempting to combine '
-                                                 + 'sections with different '
-                                                 + 'timebases',
-                                                 exception = excpt.InputError)
+            assrt.equal_arrays(*timeBases, msg='Attempting to combine '
+                               + 'sections with different '
+                               + 'timebases',
+                               exception=excpt.InputError)
         except:
             turns = len([t for t in timeBases if t == 'by_turn']) == 0
             times = len([t for t in timeBases if t == 'by_time']) == 0
```
This bit confuses me.  What's the `try/except` for?

alasheen replied:
This is for you to answer 😛 

I didn't touch anything in there except the `n_turns` in `preprocess` and the `cycle_time` part, the rest is only autopep8

scpalbright replied:
huh...
